core/src/apps/cardano/helpers/tx_builder.py

```diff
@@ -0,0 +1,188 @@
+        self.cbor_hash_builder.add_item((prev_hash, prev_index))
+
+    def finish_inputs(self) -> None:
+        self.state_machine.transition(TxBuilderStateMachine.FINISH_INPUTS)
+        self.cbor_hash_builder.finish_collection()
+
+    def start_outputs(self, outputs_count: int) -> None:
+        self.state_machine.transition(TxBuilderStateMachine.START_OUTPUTS)
+        self.cbor_hash_builder.start_collection_at_key(
+            key=TX_BODY_OUTPUTS_KEY,
+            collection=LazyCollection(LIST, outputs_count),
+        )
+
+    def add_output(self, amount: int, asset_groups_count: int, address: bytes) -> None:
+        if asset_groups_count == 0:
+            self.state_machine.transition(TxBuilderStateMachine.ADD_SIMPLE_OUTPUT)
+            self.cbor_hash_builder.add_item((address, amount))
+        else:
+            self.state_machine.transition(TxBuilderStateMachine.ADD_OUTPUT_WITH_TOKENS)
+            asset_groups = LazyCollection(DICT, asset_groups_count)
+            self.cbor_hash_builder.add_item((address, (amount, asset_groups)))
+            self.cbor_hash_builder.active_items_stack.append(asset_groups)
+
+    def finish_outputs(self) -> None:
+        self.state_machine.transition(TxBuilderStateMachine.FINISH_OUTPUTS)
+        self.cbor_hash_builder.finish_collection()
+
+    def add_asset_group(self, policy_id: bytes, tokens_count: int) -> None:
+        self.state_machine.transition(TxBuilderStateMachine.ADD_ASSET_GROUP)
+        self.cbor_hash_builder.start_collection_at_key(
+            key=policy_id,
+            collection=LazyCollection(DICT, tokens_count),
+        )
+
+    def finish_asset_groups(self) -> None:
+        self.state_machine.transition(TxBuilderStateMachine.FINISH_ASSET_GROUPS)
+        self.cbor_hash_builder.finish_collection()
+
+    def add_token(self, asset_name_bytes: bytes, amount: int) -> None:
+        self.state_machine.transition(TxBuilderStateMachine.ADD_TOKEN)
+        self.cbor_hash_builder.add_item((asset_name_bytes, amount))
+
+    def finish_tokens(self) -> None:
+        self.state_machine.transition(TxBuilderStateMachine.FINISH_TOKENS)
+        self.cbor_hash_builder.finish_collection()
+
+    def add_fee(self, fee: int) -> None:
+        self.state_machine.transition(TxBuilderStateMachine.ADD_FEE)
+        self.cbor_hash_builder.add_item((TX_BODY_FEE_KEY, fee))
+
+    def add_ttl(self, ttl: int | None) -> None:
+        if ttl is not None:
+            self.state_machine.transition(TxBuilderStateMachine.ADD_TTL)
+            self.cbor_hash_builder.add_item((TX_BODY_TTL_KEY, ttl))
+
+    def start_certificates(self, certificates_count: int) -> None:
+        self.state_machine.transition(TxBuilderStateMachine.START_CERTIFICATES)
+        self.cbor_hash_builder.start_collection_at_key(
+            key=TX_BODY_CERTIFICATES_KEY,
+            collection=LazyCollection(LIST, certificates_count),
+        )
+
+    def add_certificate(
```

Comment:
I would split this up into two functions: `add_simple_certificate` and `add_pool_registration_certificate`. The `if owners_and_relays_count is None` doesn't seem explicit enough to me and in `sign_tx.py` we can easily distinguish between which function to call.

I would then also split `owners_and_relays_count` to separate parameters.